dali/python/nvidia/dali/_utils/external_source_impl.py

```diff
@@ -255,3 +271,177 @@ def get_callback_from_source(source, cycle):
+    dtype, shape = _inspect_data(first, is_batched)
+
+    class CallableIterator:
+        first_value = first
+        def __init__(self):
+            self.source = source_desc.source
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            if CallableIterator.first_value is not None:
+                result = CallableIterator.first_value
+                CallableIterator.first_value = None
+            else:
+                result = self.source()
+            if is_batched:
+                return batch_to_numpy(result, _tf_batch_error_msg)
+            else:
+                return sample_to_numpy(result, _tf_sample_error_msg)
+
+    return CallableIterator, dtype, shape
+
+
+def get_iterable_from_iterable_or_generator(source_desc, is_batched):
+    """Wrap iterable or generator function into another iterable while peeking the first element
+
+    If the source is generator function it must be called first.
+    """
+    if source_desc.kind == SourceKind.GENERATOR_FUNC:
+        first_iter = iter(source_desc.source())
+    else:
+        first_iter = iter(source_desc.source)
+    first =  next(first_iter)
+    dtype, shape = _inspect_data(first, is_batched)
+
+    class PeekFirstGenerator:
+        first_iterator = first_iter
+        first_value = first
+        def __init__(self):
+            self.source = source_desc.source
+
+        def __iter__(self):
+            if PeekFirstGenerator.first_iterator is not None:
+                self.it = PeekFirstGenerator.first_iterator
+                PeekFirstGenerator.first_iterator = None
+            else:
+                if source_desc.kind == SourceKind.GENERATOR_FUNC:
+                    self.it = iter(source_desc.source())
+                else:
+                    self.it = iter(source_desc.source)
+            return self
+
+        def __next__(self):
+            if PeekFirstGenerator.first_value is not None:
+                result = PeekFirstGenerator.first_value
+                PeekFirstGenerator.first_value = None
+            else:
+                result = next(self.it)
+            if is_batched:
+                return batch_to_numpy(result, _tf_batch_error_msg)
+            else:
+                return sample_to_numpy(result, _tf_sample_error_msg)
+
+    return PeekFirstGenerator, dtype, shape
+
+
+def _get_generator_from_source_desc(source_desc, batch_size, is_batched):
+    """Based on DALI source description create a generator function, type and shape specification
+    compatible with TF Generator Dataset.
+
+    Cycling is delegated to the dataset as some control of some cycling behaviour cannot be
+    realized in TF.
+    """
+    if source_desc.kind == SourceKind.CALLABLE:
+        if source_desc.has_inputs:
+            if is_batched:
+                return get_batch_iterable_from_callback(source_desc)
+            else:
+                return get_sample_iterable_from_callback(source_desc, batch_size)
```

Comment:
hmmm, if is batched we do not pass the `batch_size`? Is this correct?

Reply:
In `get_batch_iterable_from_callback` we don't need the batch_size for anything, in the `get_sample_iterable_from_callback` we need to keep track of SampleInfo.